File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() < 3)
            return 0;

        int total_water = 0;
        int left = 0, right = height.size() - 1;
        int max_left = height[left], max_right = height[right];

        while (left < right) {
            if (max_left < max_right) {
                left++;
                max_left = max(max_left, height[left]);
                total_water += max_left - height[left];
            }
            else {
                right--;
                max_right = max(max_right, height[right]);
                total_water += max_right - height[right];
            }
        }

        return total_water;
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.java
class Solution {
    public int maxProfit(int[] prices) {
        int left = 0, right = 0;
        int maxProfit = 0;

        while (right < prices.length) {
            if (prices[right] < prices[left])  
                left = right;
            else 
                maxProfit = Math.max(maxProfit, prices[right] - prices[left]);

            right++;
        }

        return maxProfit;
    }
}
